experimental/sdk/webpubsub-socketio-extension/src/EIO/components/web-pubsub-transport.ts

```diff
@@ -1,28 +1,114 @@
+import { debugModule, toAsync } from "../../common/utils";
+import { ClientConnectionContext } from "./client-connection-context";
+import { WEBPUBSUB_CLIENT_CONNECTION_FILED_NAME, WEBPUBSUB_TRANSPORT_NAME } from "./constants";
 import { Transport } from "engine.io";
-import { WEBPUBSUB_TRANSPORT_NAME } from "./constants";
+import { Packet, RawData } from "engine.io-parser";
+
+const debug = debugModule("wps-sio-ext:EIO:WebPubSubTransport");
 
 /**
- * A class inherited from Engine.IO Transport class, it acts the same role with `Polling` Transport and `WebSocket` Transport.
- * Compared with the other two transports, this transport always does batch receive and optional batch send (default true).
- *
- * TODO: Batch receive is implemented in web-pubsub-translator. It should be moved here later.
+ * A class inherited from Engine.IO Transport class, it plays the same role as `Polling` Transport and `WebSocket` Transport.
+ * Similar with `Polling`, this transport always does batch send (send mulitple packets at once) and batch receive.
+ * Reference: https://github.com/socketio/engine.io/tree/6.4.2/lib/transports
  **/
 export class WebPubSubTransport extends Transport {
+  public clientConnectionContext: ClientConnectionContext;
+  /**
+   * Queue for storing packets to be sent. This design is to ensure packets are sent in order with send action in async manner.
+   */
+  public _queue: Packet[] = [];
+
+  // Reference: https://github.com/socketio/engine.io-parser/blob/5.0.7/lib/encodePacket.ts#L3
+  private _encodePacketAsync: (packet: Packet, supportsBinary: boolean) => Promise<RawData>;
+  // Reference: https://github.com/socketio/engine.io-parser/blob/5.0.7/lib/index.ts#L7
+  private _encodePayloadAsync: (packets: Packet[]) => Promise<string>;
+
+  /**
+   * Indicate whether any packet have been actually sent through AWPS or not.
+   */
+  private _onceSent: boolean;
+
+  constructor(req: unknown) {
+    debug("constructor");
+    super(req);
+    this.clientConnectionContext = req[WEBPUBSUB_CLIENT_CONNECTION_FILED_NAME];
+    this._onceSent = false;
+
+    // `Socket` places packets in its own buffer if `writable` == false. Otherwise, it calls `send` with buffer directly.
+    // Reference: https://github.com/socketio/engine.io/blob/6.4.2/lib/socket.ts#L510
+    this.writable = true;
+
+    this._encodePacketAsync = toAsync<string>(this.parser.encodePacket);
+    this._encodePayloadAsync = toAsync<string>(this.parser.encodePayload);
+  }
+
   public override supportsFraming = (): boolean => false;
 
   public override name = (): string => WEBPUBSUB_TRANSPORT_NAME;
 
-  public override async send(_packets: Packet[]): Promise<void> {
-    return Promise.resolve();
+  /**
+   * sends an array of `Packet` to the client.
+   * @param packets - An array of `Packet` to send
+   */
+  public override async send(packets: Packet[]): Promise<void> {
+    debug(`send packets, number = ${packets.length}`);
+    this.writable = false;
```

Comment:
what happens when writable is false

Reply:
Each Engine.IO Socket has an internal buffer. After `socket.flush()` is called:
1. if this.transport.writable == true, all packets in the buffer will be sent.
2. else, do nothing.  See [here](https://github.com/socketio/engine.io/blob/main/lib/socket.ts#L515). 

Before calling `socket.flush`, the socket generally has pushed new packet into the buffer. See [here](https://github.com/socketio/engine.io/blob/main/lib/socket.ts#L498)

So `writable=false` prevents `socket` sending out the buffer and clear it. During `writable=false`, the buffer length will constantly grow.